hoomd/hpmc/compute.py

```diff
@@ -7,123 +7,134 @@
-        log = analyze.log(quantities=['hpmc_free_volume'], period=100, filename='log.dat', overwrite=True)
+        mc = hoomd.hpmc.integrate.Sphere(seed=415236)
+        mc.shape["A"] = {'diameter': 1.0}
+        mc.shape["B"] = {'diameter': 0.2}
+        mc.depletant_fugacity["B"] = 1.5
+        fv = hoomd.hpmc.compute.FreeVolume(mc=mc, seed=123, test_type='B', nsample=1000)
 
     """
-    def __init__(self, mc, seed, suffix='', test_type=None, nsample=None):
-
-        # initialize base class
-        _compute.__init__(self);
-
+    def __init__(self, mc, seed, test_type=None, nsample=None):
+        # store metadata
+        param_dict = ParameterDict(
+            mc=integrate.HPMCIntegrator,
+            seed=int,
+            test_particle_type=Either([to_type_converter(str),
+                                      to_type_converter(int)]),
+            num_samples=int
+        )
+        param_dict.update(
+            dict(mc=mc,
+                 seed=int(seed),
+                 test_particle_type=str(test_type),
+                 num_samples=int(nsample)))
+        # set defaults
+        self._param_dict.update(param_dict)
+
+    def _attach(self):
+        self.test_particle_type = self._simulation.state._cpp_sys_def.getParticleData().getTypeByName(self.test_particle_type)
         # create the c++ mirror class
-        cl = _hoomd.CellList(hoomd.context.current.system_definition);
-        hoomd.context.current.system.addCompute(cl, "auto_cl3")
-
-        cls = None;
-        if not hoomd.context.current.device.cpp_exec_conf.isCUDAEnabled():
-            if isinstance(mc, integrate.sphere):
-                cls = _hpmc.ComputeFreeVolumeSphere;
-            elif isinstance(mc, integrate.convex_polygon):
-                cls = _hpmc.ComputeFreeVolumeConvexPolygon;
-            elif isinstance(mc, integrate.simple_polygon):
-                cls = _hpmc.ComputeFreeVolumeSimplePolygon;
-            elif isinstance(mc, integrate.convex_polyhedron):
-                cls = _hpmc.ComputeFreeVolumeConvexPolyhedron;
-            elif isinstance(mc, integrate.convex_spheropolyhedron):
-                cls = _hpmc.ComputeFreeVolumeSpheropolyhedron;
-            elif isinstance(mc, integrate.ellipsoid):
-                cls = _hpmc.ComputeFreeVolumeEllipsoid;
-            elif isinstance(mc, integrate.convex_spheropolygon):
-                cls =_hpmc.ComputeFreeVolumeSpheropolygon;
-            elif isinstance(mc, integrate.faceted_ellipsoid):
-                cls =_hpmc.ComputeFreeVolumeFacetedEllipsoid;
-            elif isinstance(mc, integrate.polyhedron):
-                cls =_hpmc.ComputeFreeVolumePolyhedron;
-            elif isinstance(mc, integrate.sphinx):
-                cls =_hpmc.ComputeFreeVolumeSphinx;
-            elif isinstance(mc, integrate.convex_spheropolyhedron_union):
+        cls = None
+        if isinstance(self._simulation.device, hoomd.device.CPU):
+            if isinstance(self.mc, integrate.Sphere):
```

Comment:
I would prefer using a construct like that in `hoomd/hpmc/update.py` for the `Cluster` classes `attach` method.